lib/filterx/expr-switch.c

```diff
@@ -0,0 +1,236 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "expr-switch.h"
+#include "expr-compound.h"
+#include "expr-comparison.h"
+#include "object-primitive.h"
+
+typedef struct _FilterXSwitch FilterXSwitch;
+typedef struct _FilterXSwitchCase FilterXSwitchCase;
+
+struct _FilterXSwitchCase
+{
+  FilterXUnaryOp super;
+  gsize target;
+};
+
+static inline gboolean
+filterx_switch_case_is_default(FilterXSwitchCase *self)
+{
+  return self->super.operand == NULL;
+}
+
+static void
+filterx_switch_case_set_target(FilterXSwitchCase *self, gsize target)
+{
+  self->target = target;
+}
+
+static inline gsize
+filterx_switch_case_get_target(FilterXSwitchCase *self)
+{
+  return self->target;
+}
+
+static inline FilterXObject *
+_eval_switch_case(FilterXSwitchCase *self)
+{
+  return filterx_expr_eval_typed(self->super.operand);
+}
+
+FilterXExpr *
+filterx_switch_case_new(FilterXExpr *value)
+{
+  FilterXSwitchCase *self = g_new0(FilterXSwitchCase, 1);
+  filterx_unary_op_init_instance(&self->super, FILTERX_EXPR_TYPE_NAME(switch_case), value);
+  return &self->super.super;
+}
+
+struct _FilterXSwitch
+{
+  FilterXExpr super;
+  FilterXExpr *selector;
+  FilterXObject *evaluated_selector;
+  /* pointer to FilterXSwitchCase instances */
+  GPtrArray *cases;
+  /* compound expr */
+  FilterXExpr *body;
+  gsize default_target;
+};
+
+static void
+_build_switch_table(FilterXSwitch *self, GList *body)
+{
+  self->body = filterx_compound_expr_new(FALSE);
+  for (GList *l = body; l; l = l->next)
+    {
+      FilterXExpr *expr = (FilterXExpr *) l->data;
+
+      if (filterx_expr_is_switch_case(expr))
+        {
+          FilterXSwitchCase *switch_case = (FilterXSwitchCase *) expr;
+
+          if (!filterx_switch_case_is_default(switch_case))
+            {
+              filterx_switch_case_set_target(switch_case, filterx_compound_expr_get_count(self->body));
+              g_ptr_array_add(self->cases, expr);
+            }
+          else
+            {
+              /* we do not add the default to our cases */
+              self->default_target = filterx_compound_expr_get_count(self->body);
+              filterx_expr_unref(expr);
+            }
+        }
+      else
+        filterx_compound_expr_add(self->body, expr);
+    }
+  g_list_free(body);
+}
+
+static FilterXObject *
+_eval_body(FilterXSwitch *self, gssize target)
+{
+  if (target < 0)
+    return filterx_boolean_new(TRUE);
```

Comment:
I guess this is here for the scenario where we neither have any matching cases nor have a default case. Currently I do not see any way to end up here. Shouldn't we init `self->default_target = -1;`? That would make this branch alive.

Reply:
Fixed.